include/nbl/builtin/hlsl/bxdf/transmission/beckmann.hlsl

```diff
@@ -142,25 +170,119 @@ struct SBeckmannDielectricBxDF
         const scalar_type f = fresnel::Dielectric<monochrome_type>::__call(orientedEta2, hlsl::abs<scalar_type>(params.getVdotH()))[0];
         return hlsl::promote<spectral_type>(f) * microfacet_transform;
     }
+
+    sample_type generate(NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, NBL_REF_ARG(vector3_type) u, NBL_REF_ARG(isocache_type) cache)
+    {
+        SBeckmannDielectricAnisotropicBxDF<Config> beckmann_aniso = SBeckmannDielectricAnisotropicBxDF<Config>::create(eta, A, A);
+        anisocache_type anisocache;
+        sample_type s = beckmann_aniso.generate(anisotropic_interaction_type::create(interaction), u, anisocache);
+        cache = anisocache.iso_cache;
+        return s;
+    }
+
+    scalar_type pdf(NBL_CONST_REF_ARG(params_isotropic_t) params, NBL_REF_ARG(scalar_type) onePlusLambda_V)
+    {
+        SBeckmannDG1Query<scalar_type> dg1_query;
+        dg1_query.absNdotV = params.getNdotV();
+    
+        fresnel::OrientedEtas<monochrome_type> orientedEta = fresnel::OrientedEtas<monochrome_type>::create(params.getVdotH(), hlsl::promote<monochrome_type>(eta));
+        const monochrome_type orientedEta2 = orientedEta.value * orientedEta.value;
+        dg1_query.orientedEta = orientedEta.value[0];
+
+        dg1_query.transmitted = params.cache.isTransmission();
+
+        const scalar_type reflectance = fresnel::Dielectric<monochrome_type>::__call(orientedEta2, nbl::hlsl::abs<scalar_type>(params.getVdotH()))[0];
+        
+        const scalar_type a2 = A*A;
+        ndf::Beckmann<scalar_type, false> beckmann_ndf;
+        beckmann_ndf.a2 = a2;
+        dg1_query.ndf = beckmann_ndf.template D<isocache_type>(params.cache);
+
+        dg1_query.lambda_V = beckmann_ndf.LambdaC2(params.getNdotV2());
+
+        scalar_type dg1 = beckmann_ndf.template DG1<SBeckmannDG1Query<scalar_type>, isocache_type>(dg1_query, params.cache);
+        onePlusLambda_V = dg1_query.getOnePlusLambdaV();
+
+        return hlsl::mix(reflectance, scalar_type(1.0) - reflectance, dg1_query.getTransmitted()) * dg1;
+    }
+
+    scalar_type pdf(NBL_CONST_REF_ARG(params_isotropic_t) params)
+    {
+        scalar_type dummy;
+        return pdf(params, dummy);
+    }
+
+    quotient_pdf_type quotient_and_pdf(NBL_CONST_REF_ARG(params_isotropic_t) params)
+    {
+        scalar_type onePlusLambda_V;
+        scalar_type _pdf = pdf(params, onePlusLambda_V);
+
+        ndf::Beckmann<scalar_type, false> beckmann_ndf;
+        beckmann_ndf.a2 = A*A;
+        SBeckmannG2overG1Query<scalar_type> query;
+        query.transmitted = params.getVdotH() * params.getLdotH() < 0.0;
+        query.onePlusLambda_V = onePlusLambda_V;
+        scalar_type quo = beckmann_ndf.template G2_over_G1<SBeckmannG2overG1Query<scalar_type>, sample_type>(query, params._sample);
+
+        return quotient_pdf_type::create(hlsl::promote<spectral_type>(quo), _pdf);
+    }
+
+    scalar_type A;
```

Comment:
why not store the NDF instead of its parameters, also transmission should store orientedEtas

Cook Torrance brdf and bsdf would help a lot here